compiler/plugins/target/ROCM/ROCMTargetUtils.cpp

```diff
@@ -239,38 +239,42 @@ static LogicalResult linkModuleWithGlobal(llvm::Module *module,
 }
 
 // Links ROCm-Device-Libs into the given module if the module needs it.
-void linkROCDLIfNecessary(llvm::Module *module, std::string targetChip,
-                          std::string bitCodeDir) {
-  if (!couldNeedDeviceBitcode(*module)) {
-    return;
-  }
+LogicalResult linkROCDLIfNecessary(Location loc, llvm::Module *module,
+                                   std::string targetChip,
+                                   std::string bitCodeDir) {
+  if (!couldNeedDeviceBitcode(*module))
+    return success();
   if (!succeeded(HAL::linkWithBitcodeVector(
-          module, getROCDLPaths(targetChip, bitCodeDir)))) {
-    llvm::WithColor::error(llvm::errs()) << "Fail to Link ROCDL.\n";
-  }
-  if (!succeeded(HAL::linkModuleWithGlobal(module, targetChip))) {
-    llvm::WithColor::error(llvm::errs()) << "Fail to Link with Globals.\n";
-  };
+          loc, module, getROCDLPaths(targetChip, bitCodeDir))))
+    return failure();
+  if (!succeeded(HAL::linkModuleWithGlobal(loc, module, targetChip)))
+    return failure();
+  return success();
 }
 
 // Links optimized Ukernel bitcodes into the given module if the module needs
 // it.
-void linkUkernelBCFiles(llvm::Module *module, Location loc,
-                        StringRef enabledUkernelsStr, StringRef targetChip,
-                        StringRef bitCodeDir, unsigned linkerFlags,
-                        llvm::TargetMachine &targetMachine) {
+LogicalResult linkUkernelBCFiles(Location loc, llvm::Module *module,
+                                 StringRef enabledUkernelsStr,
+                                 StringRef targetChip, StringRef bitCodeDir,
+                                 unsigned linkerFlags,
+                                 llvm::TargetMachine &targetMachine) {
   // Early exit if Ukernel not supported on target chip.
-  if (!iree_compiler::hasUkernelSupportedRocmArch(targetChip))
-    return;
+  if (!iree_compiler::hasUkernelSupportedRocmArch(targetChip)) {
+    return mlir::emitError(loc)
+           << "ukernel " << enabledUkernelsStr
+           << " not supported on target chip: " << targetChip;
```

Comment:
could wrap this string in quotes too (helps with spotting empty strings, e.g. `ukernel  not supported` vs `ukernel '' not supported`)
```suggestion
    return mlir::emitError(loc)
           << "ukernel '" << enabledUkernelsStr
           << "' not supported on target chip: " << targetChip;
```